tmpl/blog/mirage-entropy.md

```diff
@@ -0,0 +1,371 @@
+we decide to wait and try to accumulate many more events before updating the
+PRNG, we keep generating a fully predictable sequence in the meantime.
+
+And here is where Fortuna really shines. It keeps accumulating events in a
+number of
+internal pools in a round-robin fashion. These pools are constantly being
+activated, but with an exponentially decreasing frequency. The pools activated
+too frequently are wasted, but one of them is activated with just the right
+frequency to contain enough entropy to make it prohibitively expensive for an
+attacker to enumerate all the possibilities. This design was
+[shown][eat-your-entropy] to be within a constant factor from optimal entropy
+use, and in particular, scales robustly with the actual amount of entropy
+inherent in the input events.
+
+This leaves us with the problem of boot-time entropy. Not only can the saved
+random seed be reused by cloning the disk image, but in many cases, a MirageOS
+unikernel is running without such a storage at all!
+
+Following the design of [Whirlwind RNG][not-so-random], we employ an entropy
+[bootstrapping loop][bootstrap-code]. It's an iterated computation, which
+measures the time it took to perform the previous iteration, and then performs
+the amount of work that depends on the time, many times over. In this way, it
+creates a feedback loop with a fragile dependency on any non-determinism in the
+physical execution on the CPU, such as any contention or races in the CPU state.
+Even on ARM, which currently uses a less fine-grained timer and whose design is
+not as parallel as Intel's, this yields an initial value which varies wildly
+between boots. We use this value to kickstart the PRNG, giving it quick
+divergence, and ensuring that the state is unpredictable from the very start.
+
+[randomness-exposed]: http://www.cs.berkeley.edu/~cthompson/papers/vmm-entropy-report-2011.pdf
+[not-so-random]: http://www.ieee-security.org/TC/SP2014/papers/Not-So-RandomNumbersinVirtualizedLinuxandtheWhirlwindRNG.pdf
+[xentropyd]: https://github.com/mirage/xentropyd
+[eat-your-entropy]: https://eprint.iacr.org/2014/167
+[dual-ec-drgb]: https://en.wikipedia.org/wiki/Dual_EC_DRBG
+[bootstrap-code]: https://github.com/mirage/mirage-entropy/blob/863b48d4e33b43ca31c49c2e8caef4e367fab7b2/lib/entropy_xen.ml#L79
+
+
+## Parting words
+
+While some of our techniques (in particular bootstrapping on ARM) need a little
+more exposure before we place our full confidence in them -- and users should
+probably avoid generating long-term private keys in unikernels running on bare
+Xen just yet -- the combination of boostrapping, continuous reseeding, and
+robust accumulation gives us a hopefully comprehensive solution to generating
+randomness in a unikernel environment.
+
+We intend to re-evaluate the effectiveness of this design after getting some
+experience with how it works in the wild. To this end, we particularly
+appreciate the community feedback.
```

Comment:
+1 to adding mailing list subscription.

Also don't think it should be cut into two -- reads fine as it is. (It's a
good post :)

I would also prefer `timestamp` to `time stamp`...

On 9 July 2015 at 11:02, Mindy Preston <notifications@github.com> wrote:

> In tmpl/blog/mirage-entropy.md
> <https://github.com/mirage/mirage-www/pull/361#discussion_r34240421>:
>
> > +[dual-ec-drgb]: https://en.wikipedia.org/wiki/Dual_EC_DRBG
> > +[bootstrap-code]: https://github.com/mirage/mirage-entropy/blob/863b48d4e33b43ca31c49c2e8caef4e367fab7b2/lib/entropy_xen.ml#L79
> > +
> > +
> > +## Parting words
> > +
> > +While some of our techniques (in particular bootstrapping on ARM) need a little
> > +more exposure before we place our full confidence in them -- and users should
> > +probably avoid generating long-term private keys in unikernels running on bare
> > +Xen just yet -- the combination of boostrapping, continuous reseeding, and
> > +robust accumulation gives us a hopefully comprehensive solution to generating
> > +randomness in a unikernel environment.
> > +
> > +We intend to re-evaluate the effectiveness of this design after getting some
> > +experience with how it works in the wild. To this end, we particularly
> > +appreciate the community feedback.
>
> A link to the mailing list subscription page would probably not go amiss
> here, since we don't have comments locally on the page.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mirage/mirage-www/pull/361/files#r34240421>.
>



-- 
Richard Mortier
mort@cantab.net